ndk/src/midi.rs

```diff
@@ -0,0 +1,286 @@
+        self.inner.as_ptr()
+    }
+
+    /// Sends data to the specified input port.
+    pub fn send(&self, buffer: &[u8]) -> Result<usize> {
+        let num_bytes_sent = unsafe {
+            ffi::AMidiInputPort_send(self.as_ptr(), buffer.as_ptr(), buffer.len() as ffi::size_t)
+        };
+        if num_bytes_sent >= 0 {
+            Ok(num_bytes_sent as usize)
+        } else {
+            Err(
+                NdkMediaError::from_status(ffi::media_status_t(num_bytes_sent as c_int))
+                    .unwrap_err(),
+            )
+        }
+    }
+
+    /// Sends a message with a 'MIDI flush command code' to the specified port.
+    ///
+    /// This should cause a receiver to discard any pending MIDI data it may have accumulated and
+    /// not processed.
+    pub fn send_flush(&self) -> Result<()> {
+        let result = unsafe { ffi::AMidiInputPort_sendFlush(self.as_ptr()) };
+        NdkMediaError::from_status(result)
+    }
+
+    pub fn send_with_timestamp(&self, buffer: &[u8], timestamp: i64) -> Result<usize> {
+        let num_bytes_sent = unsafe {
+            ffi::AMidiInputPort_sendWithTimestamp(
+                self.as_ptr(),
+                buffer.as_ptr(),
+                buffer.len() as ffi::size_t,
+                timestamp,
+            )
+        };
+        if num_bytes_sent >= 0 {
+            Ok(num_bytes_sent as usize)
+        } else {
+            Err(
+                NdkMediaError::from_status(ffi::media_status_t(num_bytes_sent as c_int))
+                    .unwrap_err(),
+            )
+        }
+    }
+}
+
+impl<'a> Drop for MidiInputPort<'a> {
+    fn drop(&mut self) {
+        unsafe { ffi::AMidiInputPort_close(self.as_ptr()) };
+    }
+}
+
+pub struct MidiOutputPort<'a> {
+    inner: NonNull<ffi::AMidiOutputPort>,
+    _marker: PhantomData<&'a MidiDevice>,
+}
+
+impl<'a> fmt::Debug for MidiOutputPort<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("MidiOutputPort")
+            .field("inner", &self.inner)
+            .finish()
+    }
+}
+
+impl<'a> MidiOutputPort<'a> {
+    /// Creates an `MidiOutputPort` from a pointer.
+    ///
+    /// # Safety
+    /// By calling this function, you assert that the pointer is a valid pointer to a native
+    /// `AMidiOutputPort`.
+    pub unsafe fn from_ptr(inner: NonNull<ffi::AMidiOutputPort>) -> Self {
+        Self {
+            inner,
+            _marker: PhantomData,
+        }
+    }
+
+    fn as_ptr(&self) -> *mut ffi::AMidiOutputPort {
+        self.inner.as_ptr()
+    }
+
+    /// Receives the next pending MIDI message.
+    ///
+    /// To retrieve all pending messages, the client should repeatedly call this method until it
+    /// returns [`Ok(MidiOpcode::NoMessage)`].
+    ///
+    /// Note that this is a non-blocking call. If there are no Midi messages are available, the
+    /// function returns [`Ok(MidiOpcode::NoMessage)`] immediately (for 0 messages received).
```

Comment:
Nit: this returns a tuple with a `timestamp`.

Is that timestamp still valid if there was `NoMessage`? If so, I think the timestamp should be moved into the `MidiOpcode` variants where it is considered valid/useful instead.

Reply:
I have never used MIDI with timestamps - I will run a test and make a commit as you suggested.

Reply:
I haven't run the test yet but seems like it's not. 
https://cs.android.com/android/platform/superproject/+/master:frameworks/base/media/native/midi/amidi.cpp;l=344;drc=89b01aeaa9e19377ff547baab791277719b8aaf3?q=AMidi&ss=android%2Fplatform%2Fsuperproject
![image](https://user-images.githubusercontent.com/17005454/194037209-007f0952-351c-438d-af2c-27ebf48b520f.png)

Reply:
Neat, let's not expose it to the user then :)

Reply:
![image](https://user-images.githubusercontent.com/17005454/194039520-2a3670bd-2512-47b4-bf41-bbab6de346f5.png)
![image](https://user-images.githubusercontent.com/17005454/194039640-08af3593-3096-43c6-89d7-83fae1a1d98c.png)
Fixed!